File: KeepMeter/Services/DecisionEngine.swift
import Foundation

enum DecisionStatus: String, CaseIterable {
    case keep
    case review
    case returnCandidate
}

enum DecisionReason: String, CaseIterable {
    case deadlinePassed
    case unusedUrgent
    case lightlyUsedUrgent
    case unusedLateWindow
    case repeatedUse
    case earlyWindow
    case needsMoreSignal
}

struct DecisionSnapshot {
    let status: DecisionStatus
    let reason: DecisionReason
    let daysRemaining: Int
    let useCount: Int
    let costPerUse: Double?
}

enum DecisionEngine {
    static func evaluate(_ purchase: Purchase, referenceDate: Date = .now) -> DecisionSnapshot {
        let daysRemaining = purchase.daysRemaining(referenceDate: referenceDate)
        let useCount = purchase.useCount
        let elapsedRatio = purchase.returnWindowElapsedRatio(referenceDate: referenceDate)

        let status: DecisionStatus
        let reason: DecisionReason

        if daysRemaining < 0 {
            status = .review
            reason = .deadlinePassed
        } else if useCount == 0 && daysRemaining <= 3 {
            status = .returnCandidate
            reason = .unusedUrgent
        } else if useCount <= 1 && daysRemaining <= 3 {
            status = .review
            reason = .lightlyUsedUrgent
        } else if useCount == 0 && elapsedRatio >= 0.60 {
            status = .review
            reason = .unusedLateWindow
        } else if useCount >= 3 {
            status = .keep
            reason = .repeatedUse
        } else if elapsedRatio < 0.35 {
            status = .review
            reason = .earlyWindow
        } else {
            status = .review
            reason = .needsMoreSignal
        }

        return DecisionSnapshot(
            status: status,
            reason: reason,
            daysRemaining: daysRemaining,
            useCount: useCount,
            costPerUse: purchase.costPerUse
        )
    }
}

/// Single source of truth for v1 Free/Pro access rules.
///
/// `activePurchaseCount` must only include purchases whose outcome is `.active`.
/// Completed/archived purchases therefore never consume a Free slot.
enum AccessPolicy {
    static let freeActivePurchaseLimit = 5

    static func canAddActivePurchase(activePurchaseCount: Int, isPro: Bool) -> Bool {
        isPro || activePurchaseCount < freeActivePurchaseLimit
    }

    static func hasReachedFreeLimit(activePurchaseCount: Int, isPro: Bool) -> Bool {
        !canAddActivePurchase(activePurchaseCount: activePurchaseCount, isPro: isPro)
    }
}

File: android/app/src/main/java/de/kamilunavo/keepmeter/domain/DecisionEngine.kt
package de.kamilunavo.keepmeter.domain

import java.time.Instant
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import kotlin.math.max
import kotlin.math.min

enum class PurchaseOutcome { ACTIVE, KEPT, RETURNED }
enum class DecisionStatus { KEEP, REVIEW, RETURN_CANDIDATE }
enum class DecisionReason {
    DEADLINE_PASSED,
    UNUSED_URGENT,
    LIGHTLY_USED_URGENT,
    UNUSED_LATE_WINDOW,
    REPEATED_USE,
    EARLY_WINDOW,
    NEEDS_MORE_SIGNAL,
}

data class PurchaseSnapshot(
    val price: Double,
    val purchaseDateEpochMillis: Long,
    val returnDeadlineEpochMillis: Long,
    val usageTimestampsEpochMillis: List<Long>,
)

data class DecisionSnapshot(
    val status: DecisionStatus,
    val reason: DecisionReason,
    val daysRemaining: Int,
    val useCount: Int,
    val costPerUse: Double?,
)

object DecisionEngine {
    fun evaluate(
        purchase: PurchaseSnapshot,
        referenceEpochMillis: Long = System.currentTimeMillis(),
        zoneId: ZoneId = ZoneId.systemDefault(),
    ): DecisionSnapshot {
        val referenceDate = Instant.ofEpochMilli(referenceEpochMillis).atZone(zoneId).toLocalDate()
        val deadlineDate = Instant.ofEpochMilli(purchase.returnDeadlineEpochMillis).atZone(zoneId).toLocalDate()
        val daysRemaining = ChronoUnit.DAYS.between(referenceDate, deadlineDate).toInt()
        val useCount = purchase.usageTimestampsEpochMillis.size

        val total = purchase.returnDeadlineEpochMillis - purchase.purchaseDateEpochMillis
        val elapsed = referenceEpochMillis - purchase.purchaseDateEpochMillis
        val elapsedRatio = if (total > 0) min(max(elapsed.toDouble() / total.toDouble(), 0.0), 1.0) else 1.0

        val (status, reason) = when {
            daysRemaining < 0 -> DecisionStatus.REVIEW to DecisionReason.DEADLINE_PASSED
            useCount == 0 && daysRemaining <= 3 -> DecisionStatus.RETURN_CANDIDATE to DecisionReason.UNUSED_URGENT
            useCount <= 1 && daysRemaining <= 3 -> DecisionStatus.REVIEW to DecisionReason.LIGHTLY_USED_URGENT
            useCount == 0 && elapsedRatio >= 0.60 -> DecisionStatus.REVIEW to DecisionReason.UNUSED_LATE_WINDOW
            useCount >= 3 -> DecisionStatus.KEEP to DecisionReason.REPEATED_USE
            elapsedRatio < 0.35 -> DecisionStatus.REVIEW to DecisionReason.EARLY_WINDOW
            else -> DecisionStatus.REVIEW to DecisionReason.NEEDS_MORE_SIGNAL
        }

        return DecisionSnapshot(
            status = status,
            reason = reason,
            daysRemaining = daysRemaining,
            useCount = useCount,
            costPerUse = if (useCount > 0) purchase.price / useCount else null,
        )
    }
}

object AccessPolicy {
    const val FREE_ACTIVE_PURCHASE_LIMIT = 5

    fun canAddActivePurchase(activePurchaseCount: Int, isPro: Boolean): Boolean =
        isPro || activePurchaseCount < FREE_ACTIVE_PURCHASE_LIMIT

    fun hasReachedFreeLimit(activePurchaseCount: Int, isPro: Boolean): Boolean =
        !canAddActivePurchase(activePurchaseCount, isPro)
}
